### PR TITLE
feat: allow heterogeneous fits [DET-8800]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
Allow jobs to fit on to different size agents.

## Test Plan
- [x] tested manually with artificial slots, added unit test.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by NicholasBlaskey — COMMENTED** [review]
looks pretty good so far

still trying to understand the ```dedicatedAgentFits``` will approve / comment in a bit

File: master/internal/rm/agentrm/fitting.go [hunk 1]
```diff
@@ -150,25 +165,52 @@ func findDedicatedAgentFits(
...
-		c.Slots = slotsPerContainer
+	// Then try to find a fit among agents with heterogeneous sizes.
+	if allowHeterogeneousAgentFits {
+		var fits []*fittingState
+
+		// Progressively fit against nodes from largest to smallest. This works when node sizes are
+		// powers of some k, since a consuming some size k**2 is equivalent to consuming k nodes of
+		// k size, but with more nodes left find a fit. But if sizes aren't powers of some k, it
+		// doesn't work; with sizes 4, 6, and 8, a 12 slot job won't fit since we don't backtrack
+		// after not finding a fit with 8 and 6.
```
Comment (NicholasBlaskey) [inline master/internal/rm/agentrm/fitting.go:197]: do we have some way to enforce this / warn if this assumption is violated?

File: master/internal/rm/agentrm/fitting.go [hunk 2]
```diff
@@ -74,7 +82,7 @@ func findFits(
 	if req.FittingRequirements.SingleAgent || req.SlotsNeeded <= 1 {
 		return nil
 	}
-	if fits := findDedicatedAgentFits(req, agents, fittingMethod); len(fits) != 0 {
+	if fits := findDedicatedAgentFits(req, agents, fittingMethod, true); len(fits) != 0 {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/rm/agentrm/fitting.go:85]: oops, good catch

File: master/internal/rm/agentrm/fitting.go [hunk 3]
```diff
@@ -93,7 +101,7 @@ func isViable(
 
 func findDedicatedAgentFits(
 	req *sproto.AllocateRequest, agentStates map[*actor.Ref]*agentState,
-	fittingMethod SoftConstraint,
+	fittingMethod SoftConstraint, allowHeterogeneousAgentFits bool,
 ) []*fittingState {
 	if len(agentStates) == 0 {
 		return nil
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/rm/agentrm/fitting.go:107]: done

File: master/internal/rm/agentrm/fitting_test.go [hunk 4]
```diff
@@ -405,6 +406,41 @@ func TestFindDedicatedAgentFits(t *testing.T) {
 				SingleAgent: false,
 			},
 		},
+		{
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/rm/agentrm/fitting_test.go:409]: sure

**Review by NicholasBlaskey — APPROVED** [review]
LGTM

File: master/internal/rm/agentrm/fitting.go (hunk 1 above)
Comment (stoksc) [reply master/internal/rm/agentrm/fitting.go:197]: eh, maybe not a bad idea, but for now it is just in the docs (i doubt anyone would see a warning buried in master logs, anyway)

**Pull request merged**